tests/functional/test_search.py

```diff
@@ -107,6 +108,8 @@ def test_run_method_should_return_success_when_find_packages():
         options, args = command.parse_args(cmdline.split())
         status = command.run(options, args)
     assert status == SUCCESS
+    # Avoid hitting PyPi rate limits
+    time.sleep(2)
```

Comment:
I don't exactly like this approach but it's probably okay while trying to make things work initially.

Reply:
Ouch. That is ugly but I can understand its practicability. I wonder if we can find a better alternative.

Reply:
Yea, it shouldn't make a huge impact given that all CIs seem to use `-n auto` but is still suboptimal.

Reply:
It should probably use something like requests-mock so it doesn't have to hit pypi at all

Reply:
I believe the point of these tests is to actually hit the real system on the other end.

Reply:
Yup. These are intentionally written with minimal mocking, to ensure we work with "today's" PyPI.

Reply:
I'm sure there's a long story of how pypi broke something that lead to it :)  One thing I'd note about OpenDev is that the images are all the same, but unlike CI systems that only have one provider, Opendev jobs run in very heterogeneous clouds.  For networking, some have ipv4 only, some have ipv6 only with ipv4 via NAT and the nodes are spread over the northern hemisphere.

This means that tests like this often become more about things like internet connectivity, cloud-provider network config or things like upstream rate-limiting; rather than actually testing the bits of code you're interested in.  This is why we preconfigure fairly extensive cloud-region local mirrors; you find out just how unreliable the internet is pretty quickly when you're running jobs 24x7 :)

This environment does tend to act as a spotlight for such issues; things that users often hit but often write off as a "flaky" tests.

Reply:
Technically, we could attempt spawning own warehouse instance for tests but it'd need some effort setting up more instrumentation around the integration testing.

Reply:
FWIW, all our network tests automatically get marked as "flaky" and actually have to fail 5 times to be marked as an actual failure.

Reply:
Honestly, I'm fine with the time.sleep here, but I suggest you correctly capitalize PyPI. ;)

Reply:
AH! I think I see what's happening here! If you export a `CI` variable in the environment, the tests will be more robust to these failures (the flaky marking logic is conditional to that environment variable). That'll also remove the need for this time.sleep (since it's actually not useful to have it, given that the rate-limiting is IP-based and there are multiple threads which could each try at different times.

Reply:
> AH! I think I see what's happening here! If you export a CI variable in the environment, the tests will be more robust to these failures (the flaky marking logic is conditional to that environment variable). 

Ok, that seems to show some problematic tests being re-run, I've removed the sleeps; e.g.

https://storage.gra.cloud.ovh.net/v1/AUTH_dcaab5e32b234d56b626f72581e3644c/zuul_opendev_logs_9f9/9107/34cdad4d80576448c044d4ccd0f7e7b499d5427a/check/pypa-pip-py38/9f96c7a/test-results.html